jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Query.java

```diff
@@ -241,6 +276,48 @@ public String makeTypeNameValueString(List<String> typeNames, String typeNameStr
 		return ImmutableList.of();
 	}
 
+	public void subscribeToNotifications(final Server server, JMXConnection jmxConnection) throws IOException, InstanceNotFoundException {
+		MBeanServerConnection mbeanServer = jmxConnection.getMBeanServerConnection();
+		Iterable<ObjectName> queryNames = queryNames(mbeanServer);
+		for(ObjectName queryName : queryNames) {
+			// TODO: what about MBeans added / removed during runtime of the monitored application???
+			// TODO: does this whole loop make sense?? :) do we need to "query names"?
+			final ObjectInstance oi = mbeanServer.getObjectInstance(queryName);
+			Object handback = notificationProcessorFactory.create(server, this, oi);
+			NotificationListener notificationListener = new NotificationListener() {
+				@Override
+				public void handleNotification(Notification notification, Object handback) {
+					NotificationProcessor handler = (NotificationProcessor) handback;
+					handler.handleNotification(notification);
+				}
+			};
+
+			jmxConnection.addNotificationListener(oi.getObjectName(),
```

Comment:
In case of a broken connection the monitored app would need to remove the (dead) listeners. I haven't checked the JMX server code yet but there must be some logic for that.

Reply:
There isn't any logic yet. It is probably possible to plug into the connection pool to process the removal of listener before a JMXConnection is killed.

Reply:
Which connection pool are you referring to here? From our pov the monitored app is a pure JAVA app which exposes metrics via JMX. About listener removal in jmxtrans, I'm taking care of it (I hope).

Reply:
The JMX Connections are pooled. The pool is managed [by a Guice module](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/guice/JmxTransModule.java#L90-L91), and is used in [Server.execute()](https://github.com/jmxtrans/jmxtrans/blob/master/jmxtrans-core/src/main/java/com/googlecode/jmxtrans/model/Server.java#L276-L300). Having a pool for the JMX connection is mainly useful to ensure that connections are validated, recycled, and overall managed correctly. It make things a bit awkward for notifications. You can't assume that a there is a single JMX Connection to any application. You also can't assume that those connections are long lived (well, you should never assume they are, things can and will break, connections will be reset outside of your control).

Reply:
> You can't assume that a there is a single JMX Connection to any application.
That's true, but can I assume that there's a single JMX connection per configured server (might be multiple "server configs" for the same application)? If yes, my design should work.

Reply:
Nope, you can't even assume that there is a single connection per server. The [GenericKeyedObjectPool](https://commons.apache.org/proper/commons-pool/api-2.5.0/org/apache/commons/pool2/impl/GenericKeyedObjectPool.html) that we use works as a map of sub pools. This allows multiple threads to work in parallel to execute queries. Since JMXConnections are threadsafe and probably have very low contention, we could use a smarter way of managing them, getting away from the KeyedObjectPool and implementing connection validation / recycling. That might be a solution, but it requires a non trivial amount of work.